File: 01/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"time"
)

func main() {
	start := time.Now()
	partOne()
	duration := time.Since(start)
	partTwo()
	duration2 := time.Since(start)
	fmt.Printf("p1: %s, p2: %s\n", duration, duration2-duration)
}
func makeScanner(test bool) *bufio.Scanner {
	var f *os.File
	if test {
		f, _ = os.Open("inputs/testinput")
	} else {
		f, _ = os.Open("inputs/input")
	}
	reader := bufio.NewReader(f)
	return bufio.NewScanner(reader)
}
func partOne() {
	scanner := makeScanner(false)
	prev := 0
	count := 0
	for scanner.Scan() {
		line := scanner.Text()
		digit, _ := strconv.Atoi(line)
		if prev != 0 && digit > prev {
			count++
		}
		prev = digit
	}
	fmt.Println("Part 1 solution: " + strconv.Itoa(count))
}
func partTwo() {
	scanner := makeScanner(false)
	count := 0
	// I know the input size is > 3, else this is super sloppy to do
	scanner.Scan()
	l1 := scanner.Text()
	w1, _ := strconv.Atoi(l1)
	scanner.Scan()
	l2 := scanner.Text()
	w2, _ := strconv.Atoi(l2)
	scanner.Scan()
	l3 := scanner.Text()
	w3, _ := strconv.Atoi(l3)
	for scanner.Scan() {
		line := scanner.Text()
		digit, _ := strconv.Atoi(line)
		oldSum := w1 + w2 + w3
		newSum := w2 + w3 + digit

		if w2 != 0 && w3 != 0 && oldSum < newSum {
			count++
		}
		w1 = w2
		w2 = w3
		w3 = digit
	}
	fmt.Println("Part 2 solution: " + strconv.Itoa(count))
}

File: 01/main.kt
import kotlin.system.measureNanoTime
import java.io.File
import java.io.BufferedReader

fun main() {
    val part1Time = measureNanoTime {
        part1()
    }
    val part2Time = measureNanoTime {
        part2()
    }
    println("P1 in Kotlin $part1Time ns, P2 in Kotlin $part2Time ns")
}

fun part1() {
    val p1Lines = readFileAsLinesUsingBufferedReader(false)
    var prev: Int = 0
    var count: Int = 0
    for (line in p1Lines) {
        val digit = line.toIntOrNull()
        if (digit == null) {
            println("Error: failed to parse digit")
            return
        }
        if (prev != 0 && digit > prev) {
            count++;
        }
        prev = digit;
    }
    println("Part 1 solution: $count")
}

fun part2() {
    val p2Lines = readFileAsLinesUsingBufferedReader(false)
    var count: Int = 0
    var w1 = p2Lines.get(0).toInt()
    var w2 = p2Lines.get(1).toInt()
    var w3 = p2Lines.get(2).toInt()
    for (line in p2Lines.subList(3, p2Lines.size)) {
        val digit = line.toInt()
        val oldSum = w1 + w2 + w3
        val newSum = w2 + w3 + digit
        if (w2 != 0 && w3 != 0 && oldSum < newSum) {
            count++
        }
        w1 = w2
        w2 = w3
        w3 = digit
    }
    println("Part 2 solution: $count")
}

fun readFileAsLinesUsingBufferedReader(fileName: String): List<String>
    = File(fileName).bufferedReader().readLines()

fun readFileAsLinesUsingBufferedReader(useTestInput: Boolean): List<String> {
    val fileName: String = if (useTestInput) "inputs/testinput" else "inputs/input"
    return File(fileName).bufferedReader().readLines()
}